File: market-application/src/main/java/com/example/market/application/service/ExpireStaleBidsService.java
package com.example.market.application.service;

import com.example.market.application.port.in.ExpireStaleBidsUseCase;
import com.example.market.application.port.out.BidRepository;
import com.example.market.application.port.out.StaleBidScanner;
import com.example.market.domain.trading.Bid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Clock;
import java.time.Instant;
import java.util.List;

@Service
@RequiredArgsConstructor
@Slf4j
public class ExpireStaleBidsService implements ExpireStaleBidsUseCase {

    private final BidRepository bids;
    private final StaleBidScanner scanner;
    private final Clock clock;

    @Override
    @Transactional
    public int expireBatch(int batchSize) {
        Instant now = clock.instant();
        List<Bid> stale = scanner.findStaleActive(now, batchSize);
        for (Bid b : stale) {
            b.expire(now);
            bids.save(b);
        }
        if (!stale.isEmpty()) {
            log.info("expired {} bids", stale.size());
        }
        return stale.size();
    }
}

File: market-application/src/main/kotlin/com/example/market/application/service/ExpireStaleBidsService.kt
package com.example.market.application.service

import com.example.market.application.port.`in`.ExpireStaleBidsUseCase
import com.example.market.application.port.out.BidRepository
import com.example.market.application.port.out.StaleBidScanner
import java.time.Clock
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
open class ExpireStaleBidsService(
    private val bids: BidRepository,
    private val scanner: StaleBidScanner,
    private val clock: Clock,
) : ExpireStaleBidsUseCase {

    @Transactional
    override fun expireBatch(batchSize: Int): Int {
        val now = clock.instant()
        val stale = scanner.findStaleActive(now, batchSize)
        for (b in stale) {
            b.expire(now)
            bids.save(b)
        }
        if (stale.isNotEmpty()) {
            log.info("expired {} bids", stale.size)
        }
        return stale.size
    }

    companion object {
        private val log = LoggerFactory.getLogger(ExpireStaleBidsService::class.java)
    }
}
